### PR TITLE
More structured SPIR-V module builder

### PR DESCRIPTION
**Pull request by antiagainst**

### CODE REVIEW

File: tools/clang/unittests/SPIRV/StructureTest.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  auto bb = BasicBlock(4);
+  bb.addInstruction(constructInst(spv::Op::OpReturn, {}));
+  f.addBasicBlock(std::move(bb));
+  m.addFunction(std::move(f));
+  appendVector(&expected, constructInst(spv::Op::OpFunction, {1, 2, 0, 3}));
+  appendVector(&expected, constructInst(spv::Op::OpLabel, {4}));
+  appendVector(&expected, constructInst(spv::Op::OpReturn, {}));
+  appendVector(&expected, constructInst(spv::Op::OpFunctionEnd, {}));
+
+  m.setBound(6);
```
Comment (ehsannas) [inline tools/clang/unittests/SPIRV/StructureTest.cpp]: why is the bound `6` and not `5`? it looks like the last id that is used is 4, so shouldn't the bound be 5?

File: tools/clang/lib/SPIRV/Structure.cpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+void Function::clear() {
+  resultType = 0;
+  resultId = 0;
+  funcControl = spv::FunctionControlMask::MaskNone;
+  funcType = 0;
+  parameters.clear();
+  blocks.clear();
+}
+
+void Function::take(InstBuilder *builder) {
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/Structure.cpp]: Ok, i guess you don't need to assert because of line 79 below.

**Comment by ehsannas** [comment]
Good work Lei! This is shaping up nicely.

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 3]
```diff
@@ -9,31 +9,64 @@
...
-  void EndModule();
+  Status endModule();
+  /// \brief Begins building a SPIR-V function.
+  Status beginFunction(uint32_t funcType, uint32_t returnType);
+  /// \brief Ends building the current function.
+  Status endFunction();
+  /// \brief Begins building a SPIR-V basic block.
+  Status beginBasicBlock();
+  /// \brief Ends building the current SPIR-V basic block with OpReturn.
+  Status bbReturn();
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Done

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 4]
```diff
@@ -9,31 +9,64 @@
...
-  void GenHeader();
+  /// \brief Ends building the current basic block.
+  Status endBasicBlock();
 
-  SPIRVContext &TheContext;
-  /// \brief The module under building.
-  std::vector<uint32_t> TheModule;
+  SPIRVContext &theContext;                 ///< The SPIR-V context.
+  SPIRVModule theModule;                    ///< The module under building.
+  llvm::Optional<Function> theFunction;     ///< The function under buildingn.
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Done

File: tools/clang/lib/SPIRV/Structure.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+BasicBlock &BasicBlock::operator=(BasicBlock &&that) {
+  labelId = that.labelId;
+  instructions = std::move(that.instructions);
+
+  that.clear();
+
+  return *this;
+}
+
+void BasicBlock::take(InstBuilder *builder) {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/Structure.cpp]: Yep, reasonable. Done.

File: tools/clang/lib/SPIRV/Structure.cpp (hunk 2 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/Structure.cpp]: N/A

File: tools/clang/include/clang/SPIRV/Structure.h [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    llvm::Optional<uint32_t> memberIndex;
+    std::string name;
+  };
+
+  Header header; ///< SPIR-V module header.
+  std::vector<spv::Capability> capabilities;
+  std::vector<std::string> extensions;
+  std::vector<std::pair<uint32_t, std::string>> extInstSets;
+  llvm::Optional<spv::AddressingModel> addressingModel;
+  llvm::Optional<spv::MemoryModel> memoryModel;
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Structure.h]: Yeah, I was thinking about this earlier but still decided to go this way b/c

* It gives a meaningful "zero" value for addressing and memory module, so we can have a meaningful empty module.
* It makes the module a little bit more flexible.

Can definitely change this later.

File: tools/clang/unittests/SPIRV/StructureTest.cpp (hunk 1 above)
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/StructureTest.cpp]: 5 is used much more earlier at Line 112. ;P

**Pull request merged**